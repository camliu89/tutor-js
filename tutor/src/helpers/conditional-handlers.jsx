import React from 'react';
import invariant from 'invariant';
import { extend, last, first } from 'lodash';
import { Redirect } from 'react-router-dom';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.isTeacher) {
      if (course.is_concept_coach) {
        return (
          <CCDashboard courseId={courseId} {...props} />
        );
      } else {
        return (
          <Redirect
            to={{
              pathname: Router.makePathname('viewTeacherDashboard', props.params),
              query: Router.currentQuery(),
            }} />
        );
      }
    } else {
      if (course.is_concept_coach) {
        return (
          <CCStudentRedirect courseId={courseId} />
        );
      } else {
        return <StudentDashboard {...props} />;
      }
    }
  };
  // eslint-disable-next-line react/prop-types
  const legacyReferenceBookRedirect = ({ params }) => {
    const parts = params.parts.split('/');
    const course = Courses.get(first(parts));
    invariant(course, `Did not find course for params '${params}'`);
    const chapterSection = parts.length > 1 ? last(parts) : 0;
    return (
      <Redirect
        to={{
          pathname: Router.makePathname(
            chapterSection ? 'viewReferenceBookSection' : 'viewReferenceBook',
            {
              ecosystemId: course.ecosystem_id,
              chapterSection,
            }),
          query: Router.currentQuery(),
        }}
      />
    );
  };

  return {
    dashboard() { return renderDashboard; },
    legacyReferenceBookRedirect() { return legacyReferenceBookRedirect; },
  };
};

export { getConditionalHandlers };
