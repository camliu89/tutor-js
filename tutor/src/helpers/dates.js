import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { compact, reduce, min, max, map } from 'lodash';
import TimeHelper from './time';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}

export function findEarliest(dateThings) {
  return min(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}

export function findLatest(dateThings) {
  return max(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}


export function dateWithUnchangedTime(date, t) {
  if (moment.isMoment(date)) {
    date = date.format(TimeHelper.ISO_DATE_FORMAT);
  }
  return TimeHelper.getDateOnly(date);
}
