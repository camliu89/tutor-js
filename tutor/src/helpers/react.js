import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
