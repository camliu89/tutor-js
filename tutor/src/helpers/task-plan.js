import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';
import moment from 'moment';
import Time from '../models/time';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },

  calendarParams(course) {
    let date;
    const dueAtRange = course
      .teacherTaskPlans.array
      .map(plan => plan.dateRanges.due.start)
      .sort();

    if (dueAtRange.length) {
      date = dueAtRange[0];
    } else {
      date = Time.now;
    }
    return {
      to: 'calendarByDate',
      params: {
        courseId: course.id,
        date: moment(date).format('YYYY-MM-DD'),
      },
    };
  },

  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },

};
