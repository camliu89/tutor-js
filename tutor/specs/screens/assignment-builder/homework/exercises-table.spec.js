import { React, C, TimeMock, createUX } from '../helpers';
import ExercisesTable from '../../../../src/screens/assignment-builder/homework/exercises-table';

jest.mock('../../../../../shared/src/components/html', () => ({ html }) =>
  html ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null
);

describe('review exercises table', function() {
  let props;
  const now = TimeMock.setTo('2015-10-14T12:00:00.000Z');

  beforeEach(function() {
    const ux = createUX({ now });
    ux.plan.settings.exercise_ids = [ ux.exercises.array[0].id ];
    props = { ux };
  });

  it('matches snapshot', () => {
    const revex = mount(<C><ExercisesTable {...props} /></C>);
    expect.snapshot(revex.debug()).toMatchSnapshot();
    revex.unmount();
  });

  it('strips images', () => {
    const [ex] = props.ux.exercises.array;
    ex.content.questions[0].stem_html = `
      <div>
<p>hi image here: <img title="one"/></p>
<img title="two" />
        this is a test exercise
      </div>
    `;
    props.exercises = [ex];
    const et = mount(<ExercisesTable {...props} />);
    const html = et.find(`tr[data-ex-id=${ex.id}] div[dangerouslySetInnerHTML]`).props().dangerouslySetInnerHTML.__html;
    expect(html).toContain('this is a test exercise');
    expect(html).not.toContain('img');
    expect(html).toMatchSnapshot();
    et.unmount();
  });

});
