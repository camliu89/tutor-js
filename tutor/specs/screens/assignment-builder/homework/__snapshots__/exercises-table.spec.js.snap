// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`review exercises table matches snapshot 1`] = `
"<C path=\\"/\\">
  <ThemeProvider theme={{...}}>
    <MemoryRouter initialIndex={0} initialEntries={{...}}>
      <Router history={{...}}>
        <DragDropContext(Component)>
          <Component>
            <Component>
              <SpyModeWrapper>
                <Provider spyMode={{...}}>
                  <div className=\\"openstax-debug-content\\">
                    <inject-TourConductor-with-spyMode tourContext={[undefined]}>
                      <TourConductor tourContext={[undefined]} spyMode={{...}}>
                        <Provider tourContext={{...}}>
                          <div data-purpose=\\"tour-conductor-wrapper\\">
                            <TutorSpecLayout>
                              <Provider topNavbar={{...}} courseContext={{...}} bottomNavbar={{...}} setSecondaryTopControls={[Function: res]}>
                                <div>
                                  <Navbar area=\\"header\\" context={{...}} isDocked={false} />
                                  <ExerciseTable ux={{...}}>
                                    <inject-TourRegion tag=\\"table\\" delay={4000} className=\\"exercise-table\\" id=\\"add-homework-review-sections\\" courseId={1}>
                                      <TourRegion tag=\\"table\\" delay={4000} className=\\"exercise-table\\" id=\\"add-homework-review-sections\\" courseId={1} tourContext={{...}}>
                                        <table className=\\"exercise-table\\" data-tour-region-id=\\"add-homework-review-sections\\">
                                          <thead>
                                            <tr>
                                              <td />
                                              <td />
                                              <td>
                                                Problem Question
                                              </td>
                                              <td>
                                                Learning Objective
                                              </td>
                                              <td>
                                                Details
                                              </td>
                                            </tr>
                                          </thead>
                                          <tbody>
                                            <tr data-ex-id={1}>
                                              <td className=\\"exercise-number\\">
                                                1
                                              </td>
                                              <td>
                                                <ChapterSection chapterSection={{...}}>
                                                  <span className=\\"chapter-section\\" data-chapter-section=\\".0\\" />
                                                </ChapterSection>
                                              </td>
                                              <td>
                                                <Component block={false} html=\\"What conditions imply that we can use classical physics without considering special relativity or quantum mechanics?\\">
                                                  <div dangerouslySetInnerHTML={{...}} />
                                                </Component>
                                              </td>
                                              <td className=\\"ellipses\\">
                                                lo:stax-phys:1-2-1
                                              </td>
                                              <td className=\\"ellipses\\">
                                                blooms:3 / dok:3
                                              </td>
                                            </tr>
                                            <tr>
                                              <td className=\\"exercise-number\\">
                                                2
                                              </td>
                                              <td>
                                                -
                                              </td>
                                              <td>
                                                Tutor Selection
                                              </td>
                                              <td>
                                                -
                                              </td>
                                              <td>
                                                -
                                              </td>
                                            </tr>
                                            <tr>
                                              <td className=\\"exercise-number\\">
                                                3
                                              </td>
                                              <td>
                                                -
                                              </td>
                                              <td>
                                                Tutor Selection
                                              </td>
                                              <td>
                                                -
                                              </td>
                                              <td>
                                                -
                                              </td>
                                            </tr>
                                            <tr>
                                              <td className=\\"exercise-number\\">
                                                4
                                              </td>
                                              <td>
                                                -
                                              </td>
                                              <td>
                                                Tutor Selection
                                              </td>
                                              <td>
                                                -
                                              </td>
                                              <td>
                                                -
                                              </td>
                                            </tr>
                                          </tbody>
                                        </table>
                                      </TourRegion>
                                    </inject-TourRegion>
                                  </ExerciseTable>
                                  <Navbar area=\\"footer\\" context={{...}} />
                                </div>
                              </Provider>
                            </TutorSpecLayout>
                            <SpyModeContent>
                              <div className=\\"visible-when-debugging\\">
                                <div className=\\"tour-spy-info\\">
                                  available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
                                </div>
                              </div>
                            </SpyModeContent>
                          </div>
                        </Provider>
                      </TourConductor>
                    </inject-TourConductor-with-spyMode>
                    <a href=\\"#spy\\" tabIndex={-1} onClick={[Function: res]} className=\\"debug-toggle-link\\" aria-hidden=\\"true\\">
                      Ï€
                    </a>
                  </div>
                </Provider>
              </SpyModeWrapper>
            </Component>
          </Component>
        </DragDropContext(Component)>
      </Router>
    </MemoryRouter>
  </ThemeProvider>
</C>"
`;

exports[`review exercises table strips images 1`] = `
"
      <div>


        this is a test exercise
      </div>
    "
`;
